src/window_manipulation.rs

```diff
@@ -0,0 +1,374 @@
+		match self {
+			Self::Moving {
+				window,
+				orig_coords,
+				..
+			} => {
+				// If `Moving`, set the window's position to `orig_coords`.
+				debug!(
+					window = window.resource_id(),
+					orig_x = orig_coords.0,
+					orig_y = orig_coords.1,
+					"Cancelling window manipulation and undoing changes"
+				);
+				set_position(conn, window, (orig_coords.0 as i32, orig_coords.1 as i32));
+			}
+			Self::Resizing {
+				window, orig_size, ..
+			} => {
+				// If `Resizing`, set the window's dimensions to `orig_size`.
+				debug!(
+					window = window.resource_id(),
+					orig_width = orig_size.0,
+					orig_height = orig_size.1,
+					"Cancelling window manipulation and undoing changes"
+				);
+				set_dimensions(conn, window, (orig_size.0 as u32, orig_size.1 as u32));
+			}
+		}
+
+		conn.flush()?;
+		Ok(())
+	}
+
+	/// The window this window manipulation applies to.
+	pub fn window(self) -> Window {
+		match self {
+			Self::Moving { window, .. } => window,
+			Self::Resizing { window, .. } => window,
+		}
+	}
+
+	/// The starting position of the cursor when the window manipulation begins.
+	///
+	/// See also: [`cursor_x()`](WindowManipulation::cursor_x),
+	/// [`cursor_y()`](WindowManipulation::cursor_y)
+	pub fn cursor_pos(self) -> (i16, i16) {
+		match self {
+			Self::Moving { cursor_pos, .. } => cursor_pos,
+			Self::Resizing { cursor_pos, .. } => cursor_pos,
+		}
+	}
+
+	/// The x coordinate of the cursor when the window manipulation begins.
+	///
+	/// See also: [`cursor_pos()`](WindowManipulation::cursor_pos),
+	/// [`cursor_y()`](WindowManipulation::cursor_y)
+	pub fn cursor_x(self) -> i16 {
+		self.cursor_pos().0
+	}
+
+	/// The y coordinate of the cursor when the window manipulation begins.
+	///
+	/// See also: [`cursor_pos()`](WindowManipulation::cursor_pos),
+	/// [`cursor_y()`](WindowManipulation::cursor_y)
+	pub fn cursor_y(self) -> i16 {
+		self.cursor_pos().1
+	}
+
+	/// The difference between the given cursor position and the starting
+	/// [`cursor_pos()`](WindowManipulation::cursor_pos).
+	///
+	/// See also: [`diff_x(x: i16)`], [`diff_y(y: i16)`]
+	pub fn diff(self, cursor_pos: (i16, i16)) -> (i16, i16) {
+		(self.diff_x(cursor_pos.0), self.diff_y(cursor_pos.1))
+	}
+
+	/// The difference between the given x coordinate and the starting
+	/// [`cursor_x()`](WindowManipulation::cursor_x).
+	///
+	/// See also: [`diff(cursor_pos: (i16, i16))`](WindowManipulation::diff),
+	/// [`diff_y(y: i16)`](WindowManipulation::diff_y)
+	pub fn diff_x(self, x: i16) -> i16 {
+		x - self.cursor_pos().0
+	}
+
+	/// The difference between the given y coordinate and the starting
+	/// [`cursor_y()`](WindowManipulation::cursor_y).
+	///
+	/// See also: [`diff(cursor_pos: (i16, i16))`](WindowManipulation::diff),
+	/// [`diff_x(x: i16)`](WindowManipulation::diff_x)
+	pub fn diff_y(self, y: i16) -> i16 {
+		y - self.cursor_pos().1
+	}
+
+	/// Applies the window manipulation to the given window.
```

Comment:
Might want to expand the doc comment here to cover how the manipulation is applied, and the exact effects of that (and the specifics for both moving and resizing)?